metals/src/main/scala/scala/meta/internal/metals/BuildTargetClasses.scala

```diff
@@ -0,0 +1,82 @@
+package scala.meta.internal.metals
+
+import java.util
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicReference
+import ch.epfl.scala.bsp4j.ScalaMainClass
+import ch.epfl.scala.{bsp4j => b}
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.meta.internal.metals.MetalsEnrichments._
+
+final class BuildTargetClasses(
+    buildServer: () => Option[BuildServerConnection]
+)(implicit val ec: ExecutionContext) {
+  private val buildTargets =
+    new ConcurrentHashMap[b.BuildTargetIdentifier, Classes]()
+
+  def isInitialized(target: b.BuildTargetIdentifier): Boolean =
+    classesOf(target).initialized.get()
+
+  def main(target: b.BuildTargetIdentifier): util.Map[String, ScalaMainClass] =
+    classesOf(target).main
+
+  val onCompiled: BatchedFunction[b.BuildTargetIdentifier, Unit] =
+    BatchedFunction.fromFuture(fetchMainClassesFor)
+
+  private def fetchMainClassesFor(
+      targets: Seq[b.BuildTargetIdentifier]
+  ): Future[Unit] = {
+    for (target <- targets) {
+      classesOf(target).clear()
+    }
+
+    buildServer() match {
+      case Some(connection) =>
+        val parameters = new b.ScalaMainClassesParams(targets.asJava)
+        val task = for {
+          mainClasses <- connection.mainClasses(parameters).asScala
+          _ = initializeMainClasses(mainClasses)
+        } yield
+          for (target <- targets) {
+            classesOf(target).initialized.set(true)
+          }
+
+        task
+      case None =>
+        Future.successful(())
+    }
+  }
+
+  private def classesOf(target: b.BuildTargetIdentifier): Classes =
+    buildTargets.computeIfAbsent(target, _ => new Classes)
+
+  private def initializeMainClasses(result: b.ScalaMainClassesResult): Unit = {
+    def createObjectSymbol(className: String): String = {
+      val isRootPackage = !className.contains(".")
```

Comment:
> `isInsideEmptyPackage`

There's a difference between the root package and empty package

- `_empty_` is the enclosing package for classes that are not defined inside a package
- `_root_` is the owner of all packages (including `_empty_`).

To work with SemanticDB symbols we normally use the following helpers
```scala
import scala.meta.internal.semanticdb.Scala._
val sym = "Toplevel."
Symbols.Global(EmptyPackage, sym) // "_empty_/Toplevel."
```

Reply:
I will use `isInsideDefaultPackage`. When I hear "it is inside an empty package" then I think: How can it be an empty package if it has something inside.
Or maybe `isInsideUnnamedPackage`? :)

Reply:
The “empty package” is defined in the Scala language spec https://www.scala-lang.org/files/archive/spec/2.12/09-top-level-definitions.html#packagings

In the Java language spec the same package is called “unnamed”.

We are dealing with SemanticDB symbols here where we use the “empty package” for both languages https://scalameta.org/docs/semanticdb/specification.html#symbol-1